### PR TITLE
Optional basic auth protection for site

### PR DESCRIPTION
**Pull request by timabell**
### Context

We want to be able to see & share next year's course listings before they are ready to be seen by candidates. This adds an optional password to protect the whole site. See the readme for usage.

### Changes proposed in this pull request

* Add a basic auth lib.
* Make it configurable based on password config.

### CODE REVIEW

File: src/Startup.cs [hunk 1]
```diff
@@ -32,12 +37,58 @@ public Startup(IConfiguration configuration, ILoggerFactory logFactory)
         public IConfiguration Configuration { get; }
 
         // This method gets called by the runtime. Use this method to add services to the container.
+
         public void ConfigureServices(IServiceCollection services)
         {
             services.AddLogging(loggingBuilder => loggingBuilder.AddSerilog(dispose: true));
 
+            if (PreLanuchMode)
```
Comment (the-daniel-rothig) [inline src/Startup.cs]: Suggestion - can you create an extension method for IServiceCollection to do this, just to keep all the code separate? e.g. `services.ProtectOptionallyWithPassword(Configuration["SITE_PASSWORD"])`

Comment (fofr) [reply src/Startup.cs]: Type: `PreLanuchMode`

Comment (timabell) [reply src/Startup.cs]: both done, good calls

**Comment by timabell** [comment]
[🔒 Secure Search and Compare UI with optional authentication](https://trello.com/c/RsrCBwvy/313-%F0%9F%94%92-secure-search-and-compare-ui-with-optional-authentication)

**Pull request merged**